packages/next/src/client/use-merged-ref.ts

```diff
@@ -0,0 +1,46 @@
+import { useMemo, type Ref } from 'react'
+
+export function useMergedRef<TElement>(
+  refA: Ref<TElement>,
+  refB: Ref<TElement>
+): Ref<TElement> {
+  return useMemo(() => mergeRefs(refA, refB), [refA, refB])
+}
+
+export function mergeRefs<TElement>(
+  refA: Ref<TElement>,
+  refB: Ref<TElement>
+): Ref<TElement> {
+  if (!refA || !refB) {
+    return refA || refB
+  }
+
+  return (current: TElement) => {
+    const cleanupA = applyRef(refA, current)
+    const cleanupB = applyRef(refB, current)
+
+    return () => {
```

Comment:
We don't have to handle backwards compatibility with React 18 so we can always return a cleanup. If we'd have to handle React 18, we'd have to only return a cleanup when we receive a cleanup and to trigger the default React warning in 18 where you can't return a cleanup function.